### PR TITLE
feat: autocomplete mentions new ui

### LINKED ISSUE #14065: Autocomplete mentions in the input box
> As a user
> 

> I want to have my mentions to be autocompleted
> 

> So that I don’t have to type the full name
> 

### Description

When inputting `@` in the input box, I should see a list mentionable users, filtered by the text following `@` 

- If I have set a nickname it should have priority over everything and should be shown here https://www.figma.com/file/PPWkgOYlZZDxZv5SDGsZVV/Posts-%26-Attachments?node-id=2740%3A258262
- If the user has a ENS name but not a user set nickname it should be shown similarly to above
- If the user has only a display name it should be shown similarly to above
- In a one to one chat only 1 user will be in the mention box
- In a private group chat all the members will be in the mention box
- In a community all members will be in the mention box

### To be defined

- If the input box is full screen, where do we display the mention box

### Designs

https://www.figma.com/file/PPWkgOYlZZDxZv5SDGsZVV/Posts-%26-Attachments?node-id=2706%3A263716

### PR DESCRIPTION
**Pull request by OmarBasem**
fixes #14065

### Summary

This PR implements the [new UI](https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?node-id=2102%3A171080) for autocomplete mentions in chat. 

Since group chats new UI is not yet implemented, I am attaching a video showing how the mentions list looks when there are multiple suggestions.

#### Platforms

- Android
- iOS

https://user-images.githubusercontent.com/29354102/195049214-5df82b75-b904-4f9b-8eae-026c331ba199.mp4

status: ready

### CODE REVIEW

File: src/quo/design_system/colors.cljs
Comment (Parveshdhull) [inline src/quo/design_system/colors.cljs]: we shouldn't make changes in `quo/...`. Please use colors from `quo2.foundations.colors`

File: src/status_im/ui/screens/chat/views.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/views.cljs]: `autocomplete-mentions` should be a part of `chat-input-bottom-sheet` so it should be inside the component

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs]: color from quo2 should be used

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs:139]: same for color and `font-weight` should be used from quo2 foundation

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs]: why cond and not if ?

File: src/status_im/ui/screens/chat/components/style.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/style.cljs]: something seems off here

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/input.cljs]: pls use `<sub` instead of `@(re-frame/subscribe`

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/input.cljs:603]: are we sure we want to hide reply in case of suggestions ?

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/input.cljs]: could we have all calculations with conditions in `calculate-y` ? so we have only one calculate-y function ? because we 'll have more options

File: src/status_im/ui/screens/chat/styles/message/message.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/styles/message/message.cljs:179]: how pinned related to this PR ?

File: src/status_im/ui/screens/chat/views.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/views.cljs]: @flexsurfer according to the UI design I believe it shouldn't be. If the chat input is expanded, the mentions list still needs to animate up and down from the very bottom of the screen just above the keyboard. So its position needs to be independent of the chat-input.

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message.cljs]: The `color` property was already using `cond` so I thought would also use `cond` for `font-weight`. Shall I change it to `if`?

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs:603]: @flexsurfer in case of suggestions we don't need the `added-value` to `y` when replying because the bottom buttons view will not be rendered, so that will make up for the difference.

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs]: @flexsurfer `added-value` was already inside `calculate-y` in the reply component issue, but we moved outside to use it to update the input height.

File: src/status_im/ui/screens/chat/styles/message/message.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/styles/message/message.cljs:179]: Oh it's not related. I was working on the pins issue at the same time. These lines got removed anyways.

File: src/status_im/ui/screens/chat/views.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/views.cljs]: yeah i mean it still can be implemented as here, but we need to move `[rn/view` and `[components/autocomplete-mentions]` inside `chat-input-bottom-sheet` so from here it should be as one component, because currently they are related, but `splited` and used  from outside ns

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/message/message.cljs]: yeah better to use `if`

Comment (flexsurfer) [reply src/status_im/ui/screens/chat/message/message.cljs]: probably `cond` was used by mistake

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/components/input.cljs]: i mean move `calculate-y-with-mentions` code to `calculate-y` function, so we have only one

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message.cljs:136]: Didn't find any font-weight in quo2 foundation

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs]: @flexsurfer there are many calculations in both of them, and having both of them in one function maybe difficult to understand, so I created a function `get-y-value` that calls both of them.

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs]: should this be using the new text component from quo2?

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs:393]: this file is huge. Is there a clean way to break it up?

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs:374]: I guess this is where the migration strategy we discussed on Tuesday needs to start taking effect, 

If I recall correctly we want to separate old and new via the namespace, as such these additions should be in a different file.
However, I don't think we completely agreed on the how we divide these namespaces out. 
Can we make a decision so that it can be followed in this p.r (if needed)
@Parveshdhull  @flexsurfer  @ibrkhalil @ilmotta  @jo-mut

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs:646]: imo this old comment is redundant and can be removed

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs:653]: imo this old comment is redundant and should be removed 😆

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/message/message.cljs:341]: imo this old comment is also redundant

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/components/input.cljs]: I believe we have agreed to use `>evt` for re/frame dispatches so can you update this here

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (J-Son89) [inline src/status_im/ui/screens/chat/message/message.cljs]: same here, can you refactor to >evt method instead

File: src/status_im/ui/screens/chat/components/style.cljs
Comment (ibrkhalil) [inline src/status_im/ui/screens/chat/components/style.cljs]: align map

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs:393]: Thank you for your reviews @J-Son89 
There are multiple files that I would like to break up, but it will cause circular dependencies.
I think it will get smaller when we refactor the old code.

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs:646]: Why is it redundant? 🤔

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs:653]: How is it redundant? 🤨

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs]: I was wondering wouldn't it better to call the alias >dispatch since it "dispatches an event", like <sub alias "subscribes to an event"

File: src/status_im/ui/screens/chat/components/style.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/style.cljs]: Thanks!

File: src/status_im/ui/screens/chat/components/style.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/style.cljs:159]: are you sure we should have two z-indexes with value 2 ? one was 1000 before and another 2000

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/components/input.cljs:374]: i'll file a PR soon there I'll introduce new namespace for chat , i'll help with rebasing this PR

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/input.cljs]: > I was wondering wouldn't it better to call the alias >dispatch since it "dispatches an event", like <sub alias "subscribes to an event"

cc @flexsurfer

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [reply src/status_im/ui/screens/chat/components/input.cljs:646]: because if I look at the code below I can extract this information from code:
```
[quo2.button/button {:icon     true :size 32 :accessibility-label :send-message-button
                                        :on-press #(do (swap! context assoc :clear true)
                                                       (clear-input chat-id refs)
                                                       (re-frame/dispatch [:chat.ui/send-current-message]
```
e.g from the dispatch and the accessibility-label

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (J-Son89) [reply src/status_im/ui/screens/chat/components/input.cljs:653]: We can see from the code that this codeblock is related to background - e.g `styles/new-bottom-sheet-background`. So it does not really add any information that we can not get without looking at the code directly

**Review by J-Son89 — APPROVED** [review]

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/components/input.cljs]: yes i agree >evt is strange, but we should do refactoring and replace all >evt usage, not in this PR

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/input.cljs:362]: hey @OmarBasem there is no size :small and it crashes the app for some reason

**Comment by flexsurfer** [comment]
@OmarBasem please do not merge PR until it approved by dev and design teams

File: src/status_im/ui/screens/chat/components/input.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/components/input.cljs:362]: oh i see its text from quo1

File: src/status_im/ui/screens/chat/components/style.cljs
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/components/style.cljs:159]: Yes, I tested it. Why were we having 1000 and 2000 before? :D

**Comment by OmarBasem** [comment]
@flexsurfer sure, where are we doing design reviews?

**Pull request merged**
